Sources/PubNub/EventEngine/Core/Dispatcher.swift

```diff
@@ -0,0 +1,139 @@
+    self.onAnyInvocationCompleted = onAnyInvocationCompleted
+  }
+}
+
+// MARK: - Dispatcher
+
+protocol Dispatcher<Invocation, Event> {
+  associatedtype Invocation: AnyEffectInvocation
+  associatedtype Event
+      
+  func dispatch(invocations: [EffectInvocation<Invocation>], notify listener: DispatcherListener<Event>)
+}
+
+// MARK: - EffectDispatcher
+
+class EffectDispatcher<Invocation: AnyEffectInvocation, Event>: Dispatcher {
+  private let factory: any EffectHandlerFactory<Invocation, Event>
+  private let effectsCache = EffectsCache<Event>()
+  private let effectsRunner = EffectsRunner<Event>()
+    
+  init(factory: some EffectHandlerFactory<Invocation, Event>) {
+    self.factory = factory
+  }
+  
+  func hasPendingEffect(with id: String) -> Bool {
+    effectsCache.hasPendingEffect(with: id)
+  }
+    
+  func dispatch(invocations: [EffectInvocation<Invocation>], notify listener: DispatcherListener<Event>) {
+    let effectsToRun = invocations.compactMap {
+      switch $0 {
+      case .managed(let invocation):
+        return EffectWrapper(id: invocation.id, effect: factory.effect(for: invocation))
+      case .cancel(let id):
+        effectsCache.getEffect(with: id)?.cancelTask(); return nil
+      }
+    }
+    
+    effectsToRun.forEach {
+      effectsCache.put(effect: $0.effect, with: $0.id)
+    }
+    let cacheListener = DispatcherListener<Event>(onAnyInvocationCompleted: { [weak self] id, _ in
+      self?.effectsCache.removeEffect(id: id)
+    })
+    
+    effectsRunner.run(effects: effectsToRun, notify: [cacheListener, listener])
+  }
+}
+
+// MARK: - EffectsCache
+
+fileprivate class EffectsCache<Event> {
+  private var managedEffects: Atomic<[String: EffectWrapper<Event>]> = Atomic([:])
+
+  func hasPendingEffect(with id: String) -> Bool {
+    managedEffects.lockedRead { $0[id] } != nil
+  }
+  
+  func put(effect: some EffectHandler<Event>, with id: String) {
+    let existingWrapper = managedEffects.lockedRead { $0[id] }
+    existingWrapper?.effect.cancelTask()
+    managedEffects.lockedWrite { $0[id] = EffectWrapper<Event>(id: id, effect: effect) }
+  }
+    
+  func getEffect(with id: String) -> (any EffectHandler<Event>)? {
+    managedEffects.lockedRead() { $0[id] }?.effect
+  }
+  
+  func removeEffect(id: String) {
+    managedEffects.lockedWrite { $0[id] = nil }
+  }
+}
+
+// MARK: - EffectsRunner
+
+fileprivate class EffectsRunner<Event> {
+  func run(effects: [EffectWrapper<Event>], notify listeners: [DispatcherListener<Event>]) {
+    let group = DispatchGroup()
+    
+    effects.forEach { wrapper in
+      group.enter()
+      wrapper.effect.performTask(completionBlock: { (results) in
+        listeners.forEach { $0.onAnyInvocationCompleted(wrapper.id, results) }
+        group.leave()
+      })
+    }
+    group.notify(
+      queue: DispatchQueue.global(qos: .default),
+      execute: { listeners.forEach { $0.onAllInvocationsCompleted() } }
+    )
+  }
+}
```

Comment:
Do we need this runner which processes invocations in batches? Will `onAllInvocationsCompleted()` used somewhere?

Reply:
I just wanted to have a callback when all invocations did complete.
But it's not been used so far like you pointed out, so I will remove the entire `EffectsRunner`